### ISSUE TITLE
Cannot add seperator item programmatically

### ISSUE THREAD

**Issue by alashow** [vikramkakkar/SublimeNavigationView#9]
I create all navigation items programmatically, because i get data from network.

So, after some struggle, i almost achieved what i need. But i couldn't add menu seperator item.
Btw, it would be useful if you add some sample for adding menu items programmatically.

My current code:

```
private void drawNavigations(Headers headers) {
        mHeaders = headers;

        //clear added items/groups
        mMenu.clear();

        SublimeGroup mainGroup = mMenu.addGroup(false, false, true, true, SublimeGroup.CheckableBehavior.NONE);

        ArrayList<Article> articles = mHeaders.getArticles();
        for(int i = 0; i < articles.size(); i++) {
            Article article = articles.get(i);

            //if article has children, create expandable menu item
            if (article.getSubArticles().size() > 0) {
                SublimeGroup sublimeGroup = mMenu.addGroup(true, true, true, true, SublimeGroup.CheckableBehavior.NONE);

                //title of menu
                mMenu.addGroupHeaderItem(sublimeGroup.getGroupId(), article.getTitle(), article.getDescription(), false);

                //create & add new item with self path, to open parent article
                Article allArticleItem = new Article();
                allArticleItem.setTitle("All").setPath(article.getPath());
                mHeaders.getArticles().get(i).getSubArticles().add(0, allArticleItem);

                //add children
                ArrayList<Article> subArticles = article.getSubArticles();
                for(int i1 = 0; i1 < subArticles.size(); i1++) {
                    Article subArticle = subArticles.get(i1);
                    SublimeBaseMenuItem sublimeBaseMenuItem = mMenu.addTextItem(sublimeGroup.getGroupId(), subArticle.getTitle(), subArticle.getDescription(), true);
                    //set navId, to handle click
                    mHeaders.getArticles().get(i).getSubArticles().get(i1).setNavigationId(sublimeBaseMenuItem.getItemId());
                }
            } else { // else just text item
                mMenu.addTextItem(mainGroup.getGroupId(), article.getTitle(), article.getDescription(), false);
            }
        }

        SublimeGroup lastGroup = mMenu.addGroup(false, false, true, true, SublimeGroup.CheckableBehavior.NONE);
        mMenu.addSeparatorItem(lastGroup.getGroupId()); //(doesn't work)

        settingsBaseMenu = mMenu.addTextItem(lastGroup.getGroupId(), getString(R.string.preferences_title), null, true);
        settingsBaseMenu.setIcon(R.drawable.ic_settings);

        sloganView.setText(mHeaders.getSlogan());
        mMenu.finalizeUpdates();
    }
```

xml:

```
<com.appeaser.sublimenavigationviewlibrary.SublimeNavigationView
    xmlns:app="http://schemas.android.com/apk/res-auto"
    xmlns:android="http://schemas.android.com/apk/res/android"
    android:id="@+id/navigation"
    android:layout_width="304dp"
    android:layout_height="match_parent"
    android:layout_gravity="start"
    android:fitsSystemWindows="true"
    android:background="@color/white"
    app:snvHeaderLayout="@layout/navigation_header"
    app:snvMenu="@menu/navigation"/> 
```

P.S: `@menu/navigation` is empty, added so i could see preview of headerLayout (not showing if `snvMenu` not set)

Screenshot:
![](https://dotjpg.co/yLlBn.png)

I didn't add settings menu item and seperator item to xml because there is no api for adding groups before other items/groups.

Thanks.

**Comment by vikramkakkar** [comment]
> I didn't add settings menu item and seperator item to xml because there is no api for adding groups before other items/groups.

I did create an API for this:

```
// add `item` before an item with ID `pivotId`
SublimeMenu#addBefore(int pivotId, SublimebaseMenuItem item)

// add `item` after an item with ID `pivotId`
SublimeMenu#addAfter(int pivotId, SublimebaseMenuItem item)    
```

In the latest commit on `dev` branch, these methods have been modified a bit. Now you can use Group IDs to position items before/after the given Group.

**Note:** Groups are only labels that dictate how the menu items should be laid out. An empty group has no effect on the menu's visible state. So, to use a Group in the following methods, it must contain at least one item:

```
// pivotType(GROUP or ITEM) indicates whether `pivotId` identifies a Group or a menu item

SublimeMenu#addBefore(PivotType pivotType, int pivotId, SublimeBaseMenuItem item)

SublimeMenu#addAfter(PivotType pivotType, int pivotId, SublimeBaseMenuItem item)
```

To make sure that you're using the latest commit from `dev` branch, look for:

```
Current library debug version: 1
```

The version number can be verified with [Config.DEBUG_VERSION](https://github.com/vikramkakkar/SublimeNavigationView/blob/dev/sublimenavigationviewlibrary/src/main/java/com/appeaser/sublimenavigationviewlibrary/Config.java#L22).

You can look at the answers given here for help with refreshing gradle dependencies: [Link](http://stackoverflow.com/q/13565082).

> Btw, it would be useful if you add some sample for adding menu items programmatically.

Okay, I'll add info on this.

> But i couldn't add menu seperator item.

I tested your code on my end by stubbing the data models, and it works. Could you please try again with the latest commit on `dev` branch as I changed the menu-item insertion logic a bit. If the result is the same, try confirming that the separator item is indeed added by the traversing and logging the menu item list:

```
for (SublimeBaseMenuItem menuItem : mMenu.getVisibleItems()) {
    Log.i("TAG", "MenuItem type: " + menuItem.getItemType());
}
```

> P.S: @menu/navigation is empty, added so i could see preview of headerLayout (not showing if snvMenu not set)

I'll look into why this is happening.

**Comment by alashow** [comment]
I changed to dev-snapshot branch, and runned application. I didn't see any changes. So i logged the menu item list as you said, and separator item was there.

I tried changing my white background to green, and here is result:
![screenshot_20160213-191743](https://cloud.githubusercontent.com/assets/5840013/13031146/2081cab6-d287-11e5-9173-5976ea6eee4b.png)

I think it was just style issue. Sorry :) Now, waiting for #3 fix.

And about addBefore & addAfter api. I saw them before, but i couldn't use it because parameters confuses me.

If my menu.xml is:

```
<?xml version="1.0" encoding="utf-8"?>
<menu xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:app="http://schemas.android.com/apk/res-auto">

    <Separator
        android:id="@+id/separator" />

    <Text
        android:id="@+id/settings"
        android:title="Settings"/>

</menu>
```

How can I add my programmatically created groups, before @+id/separator? add(Before|After) methods last parametr is SublimebaseMenuItem, but SublimeGroup doesn't extend it.
Could you please explain this.

**Comment by vikramkakkar** [comment]
I am working on another project right now, so issue #3 will have to wait until tonight :). 

About usage of `add(Before|After)` API:

Apart from the `addTextItem(...)`, `addCheckboxItem(...)` etc. methods, `SublimeMenu` also provides the `createTextItem(...)`, `createCheckboxItem(...)` APIs to work with `add(Before|After)` methods. Here's an example:

Say you have two groups: `groupA` & `groupB`. You have defined groupB` in XML:

```
<?xml version="1.0" encoding="utf-8"?>
<menu xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:app="http://schemas.android.com/apk/res-auto">

    <SublimeGroup
        android:id="@+id/groupB" />

        <Separator
            android:id="@+id/separator" />

        <Text
            android:id="@+id/settings"
            android:title="Settings"/>

    </SublimeGroup>        

</menu>
```

Now, you wish to add `groupA` programmatically. For this example, let's say `groupA` will finally look like this:

```
<groupA>

    <Text
        android:id="@+id/someTextItem1"
        android:title="Some Text Item 1" />

    <Text
        android:id="@+id/someTextItem2"
        android:title="Some Text Item 2" />

</groupA>
```

First, we will add `groupA`. I explained in my last reply that `Groups` are only labels and cannot use `add(Before|After)` APIs. So, all we have to do right now is inform our `SublimeMenu` that a group with `ID = groupA` exists:

```
SublimeGroup groupA = mMenu.addGroup(
    false,        // isCollapsible        
    false,        // collapsed
    true,         // enabled
    true,         // visible
    SublimeGroup.CheckableBehavior.NONE    // checkableBehavior
);
```

Right now, `groupA` is empty. Let's add the first item --> `someTextItem1`

Before we can add this item, we need to create it:

```
SublimeBaseMenuItem someTextItem1 = mMenu.createTextItem(
    groupA.getGroupId(),        // groupId     
    "Some Text Item 1",          // title
    "",                                      // hint  
    false                                  // showsIconSpace
);
```

Note that we used `groupA.getGroupId()` to declare that `someTextItem1`  will belong to `groupA`. At this stage, `someTextItem1` has been created, but it _hasn't been added_ to the menu. We'll do that next.

We want that the items belonging to `groupA` should appear _before_ items in `groupB`. Here's how we can achieve this:

```
mMenu.addBefore(
    SublimeMenu.PivotType.GROUP,         // Declare that the second argument is a group id
    R.id.groupB,                                          // Provide the group id
    someTextItem1                                     // SublimeBaseMenuItem to add
);
```

In words, the above statement says: add the `SublimeBaseMenuItem` we just created -> `someTextItem1` _before_ the `SublimeGroup` that has `ID == R.id.groupB`.

The statement above is equivalent to:

```
mMenu.addBefore(
    SublimeMenu.PivotType.ITEM,             // Declare that the second argument is an item id
    R.id.separator,                                      // Provide the group id
    someTextItem1                                     // SublimeBaseMenuItem to add
);
```

Its easy enough to see why.

Final task is to add `someTextItem2`:

```
mMenu.addTextItem(groupA.getGroupId(), "Some Text Item 2", "", false);
```

One final use case. Let's say we had to add `someTextItem2` before `someTextItem1`. This is how we would do it:

```
// create & add `someTextItem2`
SublimeBaseMenuItem someTextItem2 
        = mMenu.createTextItem(groupA.getGroupId(), "Some Text Item 2", "", false);
mMenu.addBefore(SublimeMenu.PivotType.GROUP, R.id.groupB, someTextItem2);

// create & position `someTextItem1`
SublimeBaseMenuItem someTextItem1 
        = mMenu.createTextItem(groupA.getGroupId(), "Some Text Item 1", "", false);
// use `someTextItem2` to position `someTextItem1`
mMenu.addBefore(SublimeMenu.PivotType.ITEM, someTextItem2.getItemId(), someTextItem1);
```

Hope this is a bit clearer now. 

If this API is not well defined, I am open to suggestions that may help make `add(Before|After)` easier to use.

**Comment by alashow** [comment]
Yeah, now I can understand. Thanks/

About suggestions, it could be easier understand if groups are not just labels. groupHeaderItem + group. So add(Before|After) methods could be used with groups (not with first/last item of group).
But current API is understandable, after clarifying tho :)

btw, is it possible to disable icon tinting for a single menu item? In some cases menu item icons needs to be colored. Or is it "breaks" material design guidelines? 

Thanks.

**Comment by alashow** [comment]
BTW, I always get like this warnings:

```
02-16 13:49:02.646 32606-32606/package W/ResourceType: Invalid package identifier when getting bag for resource number 0xffffffff
```

**Comment by vikramkakkar** [comment]
> StateAwareTextView: textAppearance is -1

This is just for logging purposes. Its neither a warning, nor an error. Its also surrounded by the `Config.DEBUG` flag, which will be set to `false` when I put the next version on `jCenter`.

Following is the relevant code snippet from the initialization phase of [StateAwareTextView](https://github.com/vikramkakkar/SublimeNavigationView/blob/dev/sublimenavigationviewlibrary/src/main/java/com/appeaser/sublimenavigationviewlibrary/StateAwareTextView.java):

```
// If user has provided a valid value
// for @android:attr/textAppearance, it has already been
// applied in our parent - TextView. Else, resolve & use the value
// of @attr/saTextAppearance.
if (a.hasValue(R.styleable.StateAwareTextView_android_textAppearance)) {
    // Check if the value provided to `android:textAppearance` in XML is valid.
    int textAppearance = a.getResourceId(
                                              R.styleable.StateAwareTextView_android_textAppearance, -1);

    if (textAppearance == -1) { // Not valid ==> Use default value from `app:saTextAppearance`
        if (Config.DEBUG) {
            Log.i(TAG, "textAppearance is -1");
        }

        // `app:saTextAppearance` is a fallback
        int saTextAppearance = a.getResourceId(
                                                       R.styleable.StateAwareTextView_saTextAppearance,
                                                       R.style.SnvDefaultTextAppearance);

        // Version specific ways of setting TextAppearance
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            setTextAppearance(saTextAppearance);
        } else {
            setTextAppearance(getContext(), saTextAppearance);
        }
    } else {
        if (Config.DEBUG) { // `android:textAppearance` has been provided & is valid
            Log.i(TAG, "textAppearance is AVAILABLE");
        }
    }
}
```

> btw, is it possible to disable icon tinting for a single menu item? In some cases menu item icons needs to be colored. Or is it "breaks" material design guidelines?

I don't think it would break any design guidelines. But, for clarity, could you please open a new issue for this and explain the feature a bit more?

**Comment by vikramkakkar** [comment]
@alashow Issue #3 has been closed. Usage is included in the closing comment.

**Comment by sachinc19** [comment]
Hi vikramkakkar
How to set up SublimeNavigationView with SublimeNavigationView NavigationUI.setupWithNavController(navigationView, navController)
I got error:  SublimeNavigationView cannot be cast to com.google.android.material.navigation.NavigationView
please resolve my issue
